### PR TITLE
Add support for string_agg function.

### LINKED ISSUE #202: Document string_agg () function
### Related code PR

https://github.com/singularity-data/risingwave/pull/4183
https://github.com/singularity-data/risingwave/pull/3952

### Which part(s) of the docs might be affected or should be updated? And how?

https://www.risingwave.dev/docs/latest/sql-function-aggregate/
Add string_agg ().

### Reference

https://www.postgresql.org/docs/14/functions-aggregate.html
It's a common aggregate function, google it.

### PR DESCRIPTION
**Pull request by bernscode**
Add support for string_agg function. Expanded Aggregate page to include a Description column.

## Info
- **Description**: 
Add support for string_agg function.

- **Related code PR**: 
https://github.com/risingwavelabs/risingwave/pull/4183
https://github.com/risingwavelabs/risingwave/pull/3952

- **Related doc issue**: 
Resolves https://github.com/risingwavelabs/risingwave-docs/issues/202

- **Notes**: 
[ Any additional information? ]

## Before merging
  - [ ] (For version-specific PR) I have selected the corresponding software version in **Milestone** and linked the related doc issue to this PR in **Development**.
  - [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`bernscode`, `CharlieSYH`, `emile-00`, & `hengm3467`). 
  - [ ] I have checked the doc site preview and the updated parts look all good. <details><summary>How?</summary>image</details>
  - [ ] (For version-specific PR) I have confirmed that this PR is for a released version. If the software version is not released, put it on hold.

## Published doc pages
  [ After merging this PR, edit the description to include the links to the updated doc pages here. For example, https://www.risingwave.dev/docs/latest/intro/. ]

### CODE REVIEW

File: docs/sql/functions-operators/sql-function-aggregate.md [hunk 1]
```diff
@@ -5,10 +5,11 @@ title: Aggregate functions
...
+|Function|Argument Type|Return Type|Description|
+|---|---|---|---|
+|MIN()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|Minimum value of all input values.|
+|MAX()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|Maximum value of all input values.|
+|SUM()|smallint, int, bigint, numeric, real, double precision|bigint for smallint  or int arguments, numeric for bigint arguments, otherwise the same as the argument data type|Sum of all input values.|
+|COUNT()|bool, smallint, int, bigint, numeric, real, double precision, varchar|bigint|Number of input rows for which the value is not null.|
+|AVG()|	smallint, int, bigint, numeric, real, double precision|numeric for integer arguments; double precision for float point arguments| Average (arithmetic mean) of all input values|
+|STRING_AGG( expression [ORDER BY clause] ) → string|Any valid expression that can resolve to a character string. The ORDER BY clause is optional, which defines the order of concatenated string results.|string|Combines non-null values into a string.|
```
Comment (stdrc) [inline docs/sql/functions-operators/sql-function-aggregate.md]: `string_agg` actually accepts exact two arguments, the `value` to aggregate and the `delimiter`, both can be expressions of type `varchar`. The behavior is aligned with [postgres](https://www.postgresql.org/docs/current/functions-aggregate.html#:~:text=string_agg%20(%20value%20text%2C%20delimiter%20text%20)%20%E2%86%92%20text).

Also, are `DISTINCT` option and `FILTER` clause documented elsewhere? They can be optionally added in any of these agg functions.

Comment (bernscode) [reply docs/sql/functions-operators/sql-function-aggregate.md]: @richardchien I appreciate the clarification. I omitted the delimiter argument as there were no examples provided of it in the e2e tests. While I've not found any reference for the `FILTER` clause, I did find a reference for the `DISTINCT` option [on POPSQL](https://popsql.com/learn-sql/postgresql/how-to-use-stringagg-in-postgresql). I've updated the info based on your feedback. Thanks so much! Feel free to provide any additional feedback on the updates.

Comment (stdrc) [reply docs/sql/functions-operators/sql-function-aggregate.md]: See [here](https://github.com/risingwavelabs/risingwave/blob/88f1cf73f6a3e147890e8186bfbff63567eabd96/e2e_test/streaming/string_agg.slt#L17) and [here](https://github.com/risingwavelabs/risingwave/blob/88f1cf73f6a3e147890e8186bfbff63567eabd96/e2e_test/streaming/selective_agg.slt) for examples of `FILTER` clause. Also, you can check [postgres' doc](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-AGGREGATES) for more detail of the syntax. We now support the following cases:

```
aggregate_name (expression [ , ... ] [ order_by_clause ] ) [ FILTER ( WHERE filter_clause ) ]
aggregate_name (DISTINCT expression [ , ... ]) [ FILTER ( WHERE filter_clause ) ]
aggregate_name ( * ) [ FILTER ( WHERE filter_clause ) ]
```

**Note the difference, `DISTINCT` and `ORDER BY` are not supported to appear at the same time currently in our system.**

File: docs/sql/functions-operators/sql-function-aggregate.md [hunk 2]
```diff
@@ -5,10 +5,11 @@ title: Aggregate functions
...
+|Function|Argument Type|Return Type|Description|
+|---|---|---|---|
+|MIN()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|Minimum value of all input values.|
+|MAX()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|Maximum value of all input values.|
+|SUM()|smallint, int, bigint, numeric, real, double precision|bigint for smallint  or int arguments, numeric for bigint arguments, otherwise the same as the argument data type|Sum of all input values.|
+|COUNT()|bool, smallint, int, bigint, numeric, real, double precision, varchar|bigint|Number of input rows for which the value is not null.|
+|AVG()|	smallint, int, bigint, numeric, real, double precision|numeric for integer arguments; double precision for float point arguments| Average (arithmetic mean) of all input values|
+|STRING_AGG( value, delimiter ) → string|string, varchar|string|Combines non-null values into a string, separated by the delimiter. The ORDER BY clause, DISTICT option, and FILTER clause are optional.|
```
Comment (stdrc) [inline docs/sql/functions-operators/sql-function-aggregate.md]: I believe `string` is the same as `varchar`, so just leave one of them is enough.

File: docs/sql/functions-operators/sql-function-aggregate.md [hunk 3]
```diff
@@ -5,10 +5,11 @@ title: Aggregate functions
...
-|COUNT()|bool, smallint, int, bigint, numeric, real, double precision, varchar|bigint|
-|AVG()|	smallint, int, bigint, numeric, real, double precision|numeric for integer arguments; double precision for float point arguments|
+|Function|Argument Type|Return Type|Description|
+|---|---|---|---|
+|MIN()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|Minimum value of all input values.|
+|MAX()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|Maximum value of all input values.|
+|SUM()|smallint, int, bigint, numeric, real, double precision|bigint for smallint  or int arguments, numeric for bigint arguments, otherwise the same as the argument data type|Sum of all input values.|
+|COUNT()|bool, smallint, int, bigint, numeric, real, double precision, varchar|bigint|Number of input rows for which the value is not null.|
+|AVG()|	smallint, int, bigint, numeric, real, double precision|numeric for integer arguments; double precision for float point arguments| Average (arithmetic mean) of all input values|
```
Comment (CharlieSYH) [inline docs/sql/functions-operators/sql-function-aggregate.md]: ”Average (arithmetic mean) of all input values“ is missing a period at the end like the other function descriptions.

Comment (CharlieSYH) [reply docs/sql/functions-operators/sql-function-aggregate.md]: Thanks for adding descriptions to the table.

File: docs/sql/functions-operators/sql-function-aggregate.md [hunk 4]
```diff
@@ -5,11 +5,11 @@ title: Aggregate functions
...
-|STRING_AGG( value, delimiter ) → string|string, varchar|string|Combines non-null values into a string, separated by the delimiter. The ORDER BY clause, DISTICT option, and FILTER clause are optional.|
+|Function|Description|
+|---|---|
+|avg() → see description| Average (arithmetic mean) of all input values. Input types include smallint, int, bigint, numeric, real and double precision. Return type is numeric for integer inputs and double precision for float point inputs.|
+|count() → bigint|Number of input rows for which the value is not null. Input types include bool, smallint, int, bigint, numeric, real, double precision, and string.|
+|max() → same as input type|Maximum value of all input values. Input types include smallint, int, bigint, numeric, real, double precision, and string|
+|min() → same as input type|Minimum value of all input values. Input types include smallint, int, bigint, numeric, real, double precision, and string.|
+|string_agg( *value* string, delimiter ) → string|Combines non-null values into a string, separated by the delimiter. The ORDER BY clause, DISTICT option, and FILTER clause are optional.|
```
Comment (hengm3467) [inline docs/sql/functions-operators/sql-function-aggregate.md]: @bernscode DISTICT -> DISTINCT

File: docs/sql/functions-operators/sql-function-aggregate.md [hunk 5]
```diff
@@ -5,11 +5,11 @@ title: Aggregate functions
...
+|sum(smallint) → bigint <br /> sum(int) → bigint <br /> sum(bigint) → numeric <br /> sum(numeric) → same as input type <br /> sum(real) → same as input type <br /> sum(double) → same as input type <br /> sum(precision) → same as input type|Sum of all input values. Input types include smallint, int, bigint, numeric, real, and double precision. Return type is bigint for smallint  or int inputs, numeric for bigint inputs, otherwise the same as the input data type.|
```
Comment (hengm3467) [inline docs/sql/functions-operators/sql-function-aggregate.md]: @bernscode "double precision" is a single data type.

File: docs/sql/functions-operators/sql-function-aggregate.md [hunk 6]
```diff
@@ -5,10 +5,11 @@ title: Aggregate functions
...
-|MIN()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|
-|MAX()|smallint, int, bigint, numeric, real, double precision, varchar|Same as argument type|
-|SUM()|smallint, int, bigint, numeric, real, double precision|bigint for smallint  or int arguments, numeric for bigint arguments, otherwise the same as the argument data type| 
-|COUNT()|bool, smallint, int, bigint, numeric, real, double precision, varchar|bigint|
-|AVG()|	smallint, int, bigint, numeric, real, double precision|numeric for integer arguments; double precision for float point arguments|
+|Function|Description|
+|---|---|
+|avg() → see description| Average (arithmetic mean) of all input values. Input types include smallint, int, bigint, numeric, real and double precision. Return type is numeric for integer inputs and double precision for float point inputs.|
+|count() → bigint|Number of input rows for which the value is not null. Input types include bool, smallint, int, bigint, numeric, real, double precision, and string.|
+|max() → same as input type|Maximum value of all input values. Input types include smallint, int, bigint, numeric, real, double precision, and string|
```
Comment (CharlieSYH) [inline docs/sql/functions-operators/sql-function-aggregate.md]: |max() → same as input type|Maximum value of all input values. Input types include smallint, int, bigint, numeric, real, double precision, and string|

@bernscode Need a period at the end. Thanks.

**Review by CharlieSYH — APPROVED** [review]

File: docs/sql/functions-operators/sql-function-aggregate.md (hunk 1 above)
Comment (hengm3467) [reply docs/sql/functions-operators/sql-function-aggregate.md]: @richardchien and @bernscode 
FYI. I'm creating a new topic for value expressions, and will be describing aggregate expression and row constructors. Therefore, these can be de-scoped from this PR.

File: docs/sql/functions-operators/sql-function-aggregate.md (hunk 3 above)
Comment (bernscode) [reply docs/sql/functions-operators/sql-function-aggregate.md]: Nice catch! I've made this change. Thanks!

File: docs/sql/functions-operators/sql-function-aggregate.md (hunk 4 above)
Comment (bernscode) [reply docs/sql/functions-operators/sql-function-aggregate.md]: Good catch! I've made this change. Thanks!

File: docs/sql/functions-operators/sql-function-aggregate.md (hunk 6 above)
Comment (bernscode) [reply docs/sql/functions-operators/sql-function-aggregate.md]: Good catch! I've made this change. Thanks!

File: docs/sql/functions-operators/sql-function-aggregate.md (hunk 5 above)
Comment (bernscode) [reply docs/sql/functions-operators/sql-function-aggregate.md]: You're absolutely right. I've made the change. Thanks!

**Review by hengm3467 — APPROVED** [review]

**Pull request merged**